lib/mainloop.c

```diff
@@ -607,7 +608,12 @@ main_loop_run(MainLoop *self)
       cfg_load_module(self->current_configuration, "mod-python");
       debugger_start(self, self->current_configuration);
     }
-  iv_main();
+
+  if (self->options->auto_reload)
+    monitor_start(resolvedConfigurablePaths.cfgfilename, self); /* iv_main() moved inside */
```

Comment:
I would use an other approach here.
As the `monitor_start` and the `iv_main` do not really have much in common.

I would create a monitor object, which could have a call back function that triggers the reload.
Something like this:
```
if (self->options->auto_reload)
  {
    FileMonitor *fw = file_monitor_create(filename, ...);
    file_monitor_set_file_change_callback_function(fw, main_loop_reload_config, self);
    file_monitor_start(fw); //This is not a loop, just a registration of either inotify or timer
   }
...
iv_main(); //this could remain outside
```

One of the benefit of this that the `FileMonitor`  has no `MainLoop` code inside of it, it is easier to unit test also.

Reply:
> 
> 
> @Kokan - I just found a weird bug - still digging for the root cause : if you modify config file and it is not parsable, first time only config will reload (syslog-ng telling you there's a parse error) and then in successive times you try to modify it - it won't reload anymore.

@Kokan : Explanation below, any suggestions would be helpful. 
- Attempting to auto-reload a non-parsable file (for some reason - file marked by inotify as ``MOVED`` and won't be monitored anymore! 🤣 ):
```
[2018-10-24T06:01:24.031201] Finishing include; filename='/home/msalama/syslog-ng/scl/scl.conf', depth='1'
Error parsing config, root plugin asd not found in /home/msalama/install/syslog-ng/etc/syslog-ng.conf:
syslog-ng documentation: https://www.balabit.com/support/documentation?product=syslog-ng-ose
contact: https://lists.balabit.hu/mailman/listinfo/syslog-ng
[2018-10-24T06:01:24.033296] Error parsing configuration; filename='/home/msalama/install/syslog-ng/etc/syslog-ng.conf'
```
- Attempting to auto-reload a parsable file but there's a semantic issue instead of a syntax issue (no problem):

```
[2018-10-24T06:01:03.838562] Error binding socket; addr='AF_UNIX(/dev/log)', error='Address already in use (98)'
[2018-10-24T06:01:03.838891] Error initializing message pipeline; plugin_name='unix-dgram', location='source generator system:6:14'
[2018-10-24T06:01:03.839085] Error initializing new configuration, reverting to old config;
[2018-10-24T06:01:03.840744] Configuration reload finished;
```

Reply:
@Kokan - Thanks a lot for the suggestion! I've learned a lot while implementing that while the code became more compact and more traceable per function! Second commit soon. 💯

Reply:
All done! ``second_commit`` is now available for review! 😄